src/middlewared/middlewared/etc_files/smb_configure.py

```diff
@@ -0,0 +1,352 @@
+        return False
+
+    return True
+
+
+async def set_SID(middleware, config):
+    get_sid_func = "getlocalsid"
+    set_sid_func = "setlocalsid"
+
+    if config['role'] == "domain_controller":
+        get_sid_func = "getdomainsid"
+        set_sid_func = "setdomainsid"
+
+    database_SID = config['cifs']['cifs_SID']
+    system_SID = await get_system_SID(get_sid_func)
+
+    if database_SID == system_SID:
+        return True
+
+    if database_SID:
+        if not await set_system_SID(set_sid_func, database_SID):
+            logger.debug(f'Unable to set set SID to {database_SID}')
+            return False
+    else:
+        if not system_SID:
+            logger.debug(f'Unable to determine system and database SIDs')
+            return False
+
+        if not await set_database_SID(middleware, config, system_SID):
+            logger.debug(f'Unable to set database SID to {system_SID}')
+
+"""
+    Import local users into Samba's passdb.tdb file. Current behavior is to:
+    1) dump the contents of users bsdusr_smbhash to a tempory file,
+    2) import it as a legacy smbpasswd file,
+    3) remove the temporary file
+    4) disable users that are locked or have password disabled in the GUI
+
+    .users_imported sentinel file is no longer used. Instead compare the user
+    count in passdb.tdb with the list of smb users the in freenas-v1.db file.
+"""
+
+async def count_passdb_users():
+    pdb_list = await run([PDBCMD, '-d 0', '-L'])
+    if pdb_list.returncode != 0:
+        logger.debug(f'pdbedit -L command failed. Could not obtain count of users in passdb.tdb')
+        return False, 0
+
+    usercount = len(pdb_list.stdout.decode().splitlines())
+    return True, usercount
+
+
+async def write_legacy_smbpasswd(middleware, conf):
+    if not os.path.exists(TMP_PRIVATEDIR):
+        os.mkdir(TMP_PRIVATEDIR)
+        os.chmod(TMP_PRIVATEDIR, 0o700)
+
+    with open(TMP_SMBPASSWD, "w") as f:
+        for user in conf['smb_users']:
+            f.write(user['bsdusr_smbhash'] + '\n')
+
+    return True
+
+async def disable_passdb_users(smb_users):
+    for user in smb_users:
+        if user['bsdusr_locked'] or user['bsdusr_password_disabled']:
+            flags = '-d'
+        else:
+            flags = '-e'
+
+        set_pdb_flags = await run([SMBPASSWDCMD, flags, user["bsdusr_username"]], check=False)
+        if set_pdb_flags.returncode != 0:
+            logger.debug(f'failed to set flags {flags} for user {user["bsdusr_username"]} in passdb.tdb with error: {set_pdb_flags.stderr.decode()}')
+            return False
+
+    return True
+
+async def import_local_users(middleware, conf):
+    passdb_file = f"{conf['privatedir']}/passdb.tdb"
+    passdb_usercount = await count_passdb_users()
+    if len(conf['smb_users']) == passdb_usercount[1]:
+        logger.debug(f'User count in config file and passdb is identical. Bypassing import')
+        return True
+
+    ret = await write_legacy_smbpasswd(middleware, conf)
```

Comment:
write_legacy_smbpasswd does I/O operations which could eventually block middlewared if the system dataset is slow,
it would probably be safer to run in a thread

`middleware.run_in_thread(write_legacy_smbpasswd, conf)`

`write_legacy_smbpasswd` made a normal method. middleware var does not need to be passed as its not used.